// react-push-notifications@1.0.4/index.js
import React, { useState, useEffect, useRef } from 'react';

const DEFAULT_DURATION = 5000;
const DEFAULT_TEXTCOLOR = "#d2d6dc"
const DEFAULT_BGCOLOR = "#202020";
const DEFAULT_PBCOLOR = "#10b981"

export function NotifyContainer() {
    const [children, setChildren] = useState([]);

    useEffect(() => {
        window.addChild = (message, duration = DEFAULT_DURATION, textColor = DEFAULT_TEXTCOLOR, backgroundColor = DEFAULT_BGCOLOR, progressBarColor = DEFAULT_PBCOLOR) => {
            // prev.length > 0 ? prev[prev.length - 1][1] + 1 : 1;
            const id = children.length > 0 ? children[children.length - 1][1] + 1 : 1;
            setChildren(prev => [...prev, [message, id, duration, textColor, backgroundColor, progressBarColor]]);
            setTimeout(() => {
                try {
                    window.removeChild(id);
                }
                catch { }
            }, duration && duration > 0 ? duration : DEFAULT_DURATION);
        }
        window.removeChild = (index) => setChildren(prev => prev.filter(c => c[1] !== index));

        return () => {
            window.addChild = undefined;
            window.removeChild = undefined;
        };
    }, [children]);

    return (
        <div className="notify-container absolute h-screen w-full px-4 overflow-hidden select-none pointer-events-none">
            {children.map((child) => (
                <NotifyMessage key={child[1]} duration={child[2]} textColor={child[3]} backgroundColor={child[4]} progressBarColor={child[5]} onClose={() => window.removeChild(child[1])} message={child[0]} />
            ))}

        </div >
    )
};

export function notify(message, duration, textColor, bgColor, pbColor) {
    window.addChild(message, duration, textColor, bgColor, pbColor)
}

function NotifyMessage({ message, onClose, duration, textColor, backgroundColor, progressBarColor }) {
    const [progress, setProgress] = useState(1);
    const animationRef = useRef();

    useEffect(() => {
        const start = performance.now();
        const animate = (time) => {
            const elapsed = time - start;
            const progress = Math.min(1, elapsed / duration);
            setProgress(progress);
            if (progress < 1) {
                animationRef.current = requestAnimationFrame(animate);
            }
        };
        animationRef.current = requestAnimationFrame(animate);
        return () => cancelAnimationFrame(animationRef.current);
    }, [duration]);

    return (
        <div className="notify-message w-full max-w-5xl mx-auto mt-2 rounded" style={{ backgroundColor }}>
            <div className="notify-container h-12 px-4 flex justify-between items-center pointer-events-auto">
                <p style={{ color: textColor }} className="notify-message font-semibold">{message}</p>
                <button onClick={onClose} className="notify-close cursor-pointer flex justify-center items-center">
                    <svg stroke="currentColor" fill={textColor} strokeWidth="0" viewBox="0 0 512 512" height="22px" width="22px" xmlns="http://www.w3.org/2000/svg"><path d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"></path></svg>
                </button>
            </div>
            <div className="notify-progressbar w-full h-[6px]" style={{ backgroundColor: progressBarColor, width: `${100 - (progress * 100)}%` }}></div>
        </div>
    )
}

export const Colors = {
    transparent: 'transparent',
    current: 'currentColor',
    black: '#000',
    white: '#fff',
    gray: {
        50: '#f9fafb',
        100: '#f4f5f7',
        200: '#e5e7eb',
        300: '#d2d6dc',
        400: '#9fa6b2',
        500: '#6b7280',
        600: '#4b5563',
        700: '#374151',
        800: '#1f2937',
        900: '#111827',
    },
    red: {
        50: '#fef2f2',
        100: '#fee2e2',
        200: '#fecaca',
        300: '#fca5a5',
        400: '#f87171',
        500: '#ef4444',
        600: '#dc2626',
        700: '#b91c1c',
        800: '#991b1b',
        900: '#7f1d1d',
    },
    yellow: {
        50: '#fffbdd',
        100: '#fff5b1',
        200: '#ffea7f',
        300: '#ffdf5d',
        400: '#ffe250',
        500: '#ffd700',
        600: '#ffb700',
        700: '#ff8c00',
        800: '#ff6a00',
        900: '#ff4500',
    },
    green: {
        50: '#ecfdf5',
        100: '#d1fae5',
        200: '#a7f3d0',
        300: '#6ee7b7',
        400: '#34d399',
        500: '#10b981',
        600: '#059669',
        700: '#047857',
        800: '#065f46',
        900: '#064e3b',
    },
    blue: {
        50: '#eff6ff',
        100: '#dbeafe',
        200: '#bfdbfe',
        300: '#93c5fd',
        400: '#60a5fa',
        500: '#3b82f6',
        600: '#2563eb',
        700: '#1d4ed8',
        800: '#1e40af',
        900: '#1e3a8a',
    },
    indigo: {
        50: '#f0f5ff',
        100: '#e5edff',
        200: '#cddbfe',
        300: '#b4c6fc',
        400: '#8da2fb',
        500: '#6875f5',
        600: '#5850ec',
        700: '#5145cd',
        800: '#42389d',
        900: '#362f78',
    },
    purple: {
        50: '#F5F3FF',
        100: '#EDE9FE',
        200: '#DDD6FE',
        300: '#C4B5FD',
        400: '#A78BFA',
        500: '#8B5CF6',
        600: '#7C3AED',
        700: '#6D28D9',
        800: '#5B21B6',
        900: '#4C1D95',
    },
}
